### PR TITLE
Set a default ssl.peer_name context in StreamHandler

### PR DESCRIPTION
**Pull request by TimWolla**
This is required when using the `force_ip_resolve` option with the stream
handler:

As that option will cause the StreamHandler to manually resolve the hostname
and then replace the hostname with the resolved IP address in the URI, PHP
will use that IP address by default in the SNI of the TLS handshake.

Set an explicit ssl.peer_name within the stream's context based on the hostname
in the URL to fix this.

Setting a proper SNI is independent from TLS certificate validation, thus this
value must not be dependent on the `verify` option.

A test cannot be added, due to a lack of TLS support with the current testing
infrastructure. TLS support cannot easily be added, because it would require a
separate port and also certificates that would need to be commited to the
repository. However correctness can be verified by setting `force_ip_resolve`
to `v4` and attempting to make a request to `https://www.example.com/`. It will
fail without this commit and work with.

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
What if Guzzle follows a redirect to another host. Won't this break that?

**Comment by TimWolla** [comment]
It _should_ not: `getDefaultContext` already is responsible for setting the request headers (including `host` and possible authorization).

However I did not test this. If you'd like to hold off merging this, until I tested this: This is fine for me.

**Comment by GrahamCampbell** [comment]
Thanks. Let us know the outcome, please. :)

**Comment by TimWolla** [comment]
Will do, but this is not going to happen today. I will try to get around to it next week, but can't promise anything.

If you planned to release today, then please don't wait for this. I don't think this fix is super important, I just noticed this issue by accident while working on #2989.

**Comment by GrahamCampbell** [comment]
No worries, and no, we don't plan to release today. Maybe we will do releases on this repo and the psr7 repo, this time next weekend.

**Comment by TimWolla** [comment]
I've added a new TestCase that actually performs networking against third party services, it needs to be explicitly enabled by setting the GUZZLE_TEST_ALLOW_NETWORK environment variable.

With that I could confirm that redirects work as expected.

File: tests/Handler/Network/StreamHandlerTest.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \GuzzleHttp\Handler\StreamHandler
+ */
+class StreamHandlerTest extends TestCase
+{
+    public function setUp(): void
+    {
+        if (!($_ENV['GUZZLE_TEST_ALLOW_NETWORK'] ?? false)) {
```
Comment (GrahamCampbell) [inline tests/Handler/Network/StreamHandlerTest.php:19]: please use $_SERVER for reading env variables.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Handler/Network/StreamHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        $handler = new StreamHandler();
+
+        $response = $handler(
+            new Request('GET', 'https://www.example.com/'),
+            [
+                RequestOptions::STREAM => true,
+            ]
+        )->wait();
+
+        self::assertEquals(200, $response->getStatusCode());
```
Comment (GrahamCampbell) [inline tests/Handler/Network/StreamHandlerTest.php:35]: please use assertSame where possible

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Review by GrahamCampbell — APPROVED** [review]

**Comment by TimWolla** [comment]
On an unrelated note: Should the 8.0/8.1 CI be set to 'Required' in the repository settings?

**Review by GrahamCampbell — APPROVED** [review]

**Review by Nyholm — APPROVED** [review]
Thank you

**Pull request merged**